src/app/modules/submission/encrypt-submission/encrypt-submission.middleware.ts

```diff
@@ -87,3 +119,303 @@ export const validateStorageSubmissionParams = celebrate({
+  try {
+    label = 'Read file content'
+
+    const fileContentsView = new Uint8Array(attachment)
+
+    label = 'Encrypt content'
+    const encryptedAttachment = await formsgSdk.crypto.encryptFile(
+      fileContentsView,
+      publicKey,
+    )
+
+    label = 'Base64-encode encrypted content'
+    const encodedEncryptedAttachment = {
+      ...encryptedAttachment,
+      binary: encodeBase64(encryptedAttachment.binary),
+    }
+
+    return { id, encryptedFile: encodedEncryptedAttachment }
+  } catch (error) {
+    logger.error({
+      message: 'Error encrypting attachment',
+      meta: {
+        action: 'encryptAttachment',
+        label,
+        error,
+      },
+    })
+    throw error
+  }
+}
+
+const getEncryptedAttachmentsMapFromAttachmentsMap = async (
+  attachmentsMap: Record<string, Buffer>,
+  publicKey: string,
+): Promise<StorageModeAttachmentsMap> => {
+  const attachmentPromises = Object.entries(attachmentsMap).map(
+    ([id, attachment]) => encryptAttachment(attachment, { id, publicKey }),
+  )
+
+  return Promise.all(attachmentPromises).then((encryptedAttachmentsMeta) =>
+    chain(encryptedAttachmentsMeta)
+      .keyBy('id')
+      // Remove id from object.
+      .mapValues((v) => omit(v, 'id'))
+      .value(),
+  )
+}
+
+/**
+ * Encrypt submission content before saving to DB.
+ */
+export const encryptSubmission: StorageSubmissionMiddlewareHandlerType = async (
+  req,
+  res,
+  next,
+) => {
+  const formId = req.params.formId
+  const formDef = req.body.formsg.formDef
+  if (!formDef) return res.send(new FormDefinitionNotRetrievedError())
+
+  const publicKey = formDef.publicKey
+  if (!publicKey) return res.send(new FormMissingPublicKeyError())
+
+  const attachmentsMap: Record<string, Buffer> = {}
+
+  const logMeta = {
+    action: 'encryptSubmission',
+    ...createReqMeta(req),
+    formId,
+  }
+
+  // Populate attachment map
+  req.body.responses.filter(isAttachmentResponse).forEach((response) => {
+    const fieldId = response._id
+    attachmentsMap[fieldId] = response.content
+  })
+
+  const encryptedAttachments =
+    await getEncryptedAttachmentsMapFromAttachmentsMap(
+      attachmentsMap,
+      publicKey,
+    )
+
+  const filteredResponses = getFilteredResponses(formDef, req.body.responses)
+
+  if (filteredResponses.isErr()) {
+    logger.warn({
+      message: filteredResponses.error.message,
+      meta: logMeta,
+    })
+    return res.status(StatusCodes.INTERNAL_SERVER_ERROR).json({
+      message: filteredResponses.error.message,
+    })
+  }
+
+  const strippedBodyResponses = req.body.responses.map((response) => {
+    if (isAttachmentResponse(response)) {
+      return {
+        ...response,
+        filename: undefined,
+        content: undefined, //Strip out attachment content
+      }
+    } else {
+      return response
+    }
+  })
+
+  const encryptedContent = formsgSdk.crypto.encrypt(
+    strippedBodyResponses,
+    publicKey,
+  )
+
+  req.body.formsg.encryptedPayload = {
```

Comment:
likewise, should we consider putting this in res.locals?

Reply:
Discussed above in [this comment](https://github.com/opengovsg/FormSG/pull/6595#discussion_r1298074277) - conclusion is to leave it as it is 👍